### PR TITLE
Add notifications of severed relationships

### PR DESCRIPTION
**Pull request by ClearlyClaire**
When a moderation action severs relationship, record those relationships so they can be examined by the affected local users, offering them to export the lost follows, as well as re-issue them if the domain block gets reverted.

Lost *followers* cannot be recovered this way, but the information should be made available nonetheless.

image

image

TODO:
- [x] add database model for severed relationships
- [x] record severed relationships on server-wide domain suspension
- [x] record severed relationships on user domain block
- [x] record severed relationships on server-wide account suspension
- [x] allow reviewing and downloading severed relationships
- [x] add notification of severed relationships
- [x] tell when a domain is purged
- [ ] allow one-click re-issuing of severed follows

Documentation PR: https://github.com/mastodon/documentation/pull/1385

### CODE REVIEW

File: app/serializers/rest/relationship_severance_event_serializer.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+  attributes :id, :type, :purged, :target_name, :created_at
+
+  attribute :relationships_count, if: -> { current_user.present? }
+
+  def id
+    object.id.to_s
+  end
+
+  def relationships_count
+    object.severed_relationships.where(local_account: current_user.account).count
```
Comment (Gargron) [inline app/serializers/rest/relationship_severance_event_serializer.rb:13]: Potential performance bottleneck. Perhaps a counter column?

File: app/models/notification.rb [hunk 2]
```diff
@@ -37,6 +37,7 @@ class Notification < ApplicationRecord
     favourite
     poll
     update
+    relationships_severed
```
Comment (Gargron) [inline app/models/notification.rb:40]: Perhaps we could come up with a better sounding name for the API value

File: db/schema.rb [hunk 3]
```diff
@@ -814,6 +814,15 @@
     t.string "url"
   end
 
+  create_table "relationship_severance_events", force: :cascade do |t|
+    t.integer "type", null: false
+    t.string "target_name"
```
Comment (Gargron) [inline db/schema.rb:819]: We might not want null values here and the line below.

File: app/serializers/rest/relationship_severance_event_serializer.rb (hunk 1 above)
Comment (ClearlyClaire) [reply app/serializers/rest/relationship_severance_event_serializer.rb:13]: We'd need yet another table, as a relationship severance event is not about a specific account, unlike this count.

File: app/models/notification.rb (hunk 2 above)
Comment (ClearlyClaire) [reply app/models/notification.rb:40]: Do you have any suggestion?

Comment (ClearlyClaire) [reply app/models/notification.rb:40]: `severed_relationships`? `broken_relationships`?

Comment (renchap) [reply app/models/notification.rb:40]: `severed_relationships` seems consistent with the naming used elsewhere

**Review by Gargron — APPROVED** [review]

**Comment by vmstan** [comment]
Great to see this get merged!

**Comment by nanopone** [comment]
this is incredible! great job everyone

**Comment by gunchleoc** [comment]
@ClearlyClaire There are 2 similar keys in Crowdin now with the same text, is the second one a typo? `notification.severed_relationships` and `notification.severedRelationships`

Lines 41 and 378 in https://github.com/mastodon/mastodon/commit/44bf7b812894f9c8d67cbfbb1a23b0091aa601c5

**Comment by ClearlyClaire** [comment]
Oops. Thanks for pointing this out, https://github.com/mastodon/mastodon/pull/29689 should address it

**Pull request merged**